File: LeetCode/2385. Amount of Time for Binary Tree to Be Infected/Amount.py
from collections import deque

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def traversal (self,root, mp, target, startNode):
      if not root: 
        return None
      
      if(root.val == target):
        startNode[0] = root
        return root
      
      left = self.traversal(root.left,mp,target,startNode)
      if left:
        mp[left.val] = root
        return root
      
      right = self.traversal(root.right,mp,target,startNode)
      if right:
        mp[right.val] = root
        return root
      
      return None
      
  
    def amountOfTime(self, root: Optional[TreeNode], start: int) -> int:
      res = -1
      mp = {}
      startNode = [None]
      self.traversal(root,mp,start,startNode)
      q = deque(startNode[0])
      
      while q:
        size = len(q)
        res += 1 
        while size > 0:
          cur = q.popleft()
          size -= 1
          
          if cur.val in mp:
            q.append(mp[cur.val])    
          if cur.left and cur.left.val not in mp:
            q.append(cur.left)
          if cur.right and cur.right.val not in mp:
            q.append(cur.right)
            
      return res
          
          
           

File: LeetCode/2385. Amount of Time for Binary Tree to Be Infected/Amount.cpp
# include <unordered_map>
# include <queue>
using namespace std;

 // Definition for a binary tree node.
  struct TreeNode {
      int val;
      TreeNode *left;
      TreeNode *right;
      TreeNode() : val(0), left(nullptr), right(nullptr) {}
      TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
      TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
  };
 
class Solution {
public:
    // TreeNode* traversal(TreeNode* root, unordered_map<int,TreeNode*>& mp){
    //   if(!root) return nullptr;

    //   TreeNode* left = traversal(root->left,mp);
    //   if(left) {
    //     mp[root->val] = left;
    //     mp[left->val] = root;
    //   }
    //   TreeNode* right = traversal(root->right,mp);
    //   if(right) {
    //     mp[root->val] = right;
    //     mp[left->val] = root;
    //   }
    //   return root;
    // }

    TreeNode* traversal(TreeNode* root, unordered_map<int,TreeNode*>& mp ,int target, TreeNode*& startNode){
      if(!root) return nullptr;
      if(root->val == target){ 
        startNode = root;
        return root;
      }

      if(TreeNode* left = traversal(root->left,mp,target,startNode)){
        mp[left->val] = root;
        return root;
      }
      if(TreeNode* right = traversal(root->right,mp,target,startNode)){
        mp[right->val] = root;
        return root;
      }

      return nullptr;
    }

    int amountOfTime(TreeNode* root, int start) {
       int res = -1;
       unordered_map<int,TreeNode*> mp;
       TreeNode* startNode = nullptr;
       traversal(root,mp,start,startNode); 
       queue<TreeNode*> q;
       q.push(startNode);
      
       while(!q.empty()){
         int size = q.size();
         res++;
         while(size-- > 0){
           TreeNode* cur = q.front(); q.pop();
           if(mp.count(cur->val)) q.push(mp[cur->val]);
           if(cur->left && !mp.count(cur->left->val)) q.push(cur->left);
           if(cur->right && !mp.count(cur->right->val)) q.push(cur->right);
         }
       }
       return res;
    }
};